cfpb/cfgov-refresh
# New mega menu backend implementation

This commit adds a new backend implementation of the site mega menu, asa new Django app named mega_menu. This implementation can be used instead of the existing one by enabling the `MEGA_MENU_BACKEND_V2` feature flag.

This new implementation supports both the current mega menu design as well as the two variations currently being considered and tested. These two variations can be tested by enabling the `MEGA_MENU_VAR_1` or `MEGA_MENU_VAR_2` feature flags, respectively.

The major differences between the new implementation and the old one include:

1. Uses far fewer database queries. Instead of making at least one database query for each menu link pointing at a Wagtail Page, all page information is retrieved in a single query. This optimization allows us to remove a dedicated Django cache that currently exists to mitigate these queries.
2. An independent menu is now maintained for each supported language, meaning that the Spanish menu is now editable in Wagtail.
3. The data passed to the frontend is now decoupled from the database schema on the backend, hopefully allowing for greater flexibility and experimentation in future.

The new menu content can be migrated from existing menu content by running this new Django management command:

```
cfgov/manage.py migrate_menu_content
```

The new menu content can then be edited in the Wagtail admin under a new "Mega menu (v2)" section. This area contains a "Preview" button that can be used to easily test the menu without having to navigate to a page. Each menu has its own preview page, for example:

- http://localhost:8000/admin/mega_menu/menu/preview/en/
- http://localhost:8000/admin/mega_menu/menu/preview/es/

Currently the "Preview" button only shows the saved version of the menu; in future it would be nice to have this work like Wagtail page previews, and allow for previewing of unsaved edits.

## Screenshots

New Wagtail admin area. Note "Mega menu (v2)" in sidebar, and list of menus to edit, including edit, delete, and preview buttons:

![image](https://user-images.githubusercontent.com/654645/72933571-f7e5a100-3d2f-11ea-87b3-54cc354dbff1.png)

Editing a menu. Note the preview button at the bottom:

![image](https://user-images.githubusercontent.com/654645/72933630-177cc980-3d30-11ea-95a3-ce55b888628f.png)

Preview view, with placeholder content:

![image](https://user-images.githubusercontent.com/654645/72933649-23688b80-3d30-11ea-9583-b2f8a02a3e01.png)


## Testing

These testing instructions assume you are using a production dump:

1. Check out this branch, and apply its migrations to your local database using `cfgov/manage.py migrate`.
2. Migrate mega menu content from the existing tables into the new tables by running `cfgov/manage.py migrate_menu_content`.
3. In the Wagtail admin, navigate to http://localhost:8000/admin/mega_menu/menu/ and confirm that you have data for both the English and Spanish menus.
4. Try clicking the "Preview" button for both of these -- note that this still uses the existing backend and design because you haven't enabled anything new with feature flags.

Restart your local dev server with the Django Debug Toolbar enabled to observe the improvement in database queries. For example, use `ENABLE_DEBUG_TOOLBAR=1 ./runserver.sh`. Navigate back to the English preview view at http://localhost:8000/admin/mega_menu/menu/preview/en/, and, in the DJDT sidebar, you should see something like 119 queries.

5. Now, enable the new backend by going to http://localhost:8000/admin/flags/create/ and creating a flag named `MEGA_MENU_BACKEND_V2`. Enable it for all requests at http://localhost:8000/admin/flags/MEGA_MENU_BACKEND_V2/.
6. Go back to http://localhost:8000/admin/mega_menu/menu/preview/en/, and you should instead see something like 10 queries. The visual appearance and behavior of the menu should be the same.
7. Also visit the Spanish preview page at http://localhost:8000/admin/mega_menu/menu/preview/en/ and confirm that the menu looks and acts the same.
8. You can also visit the site homepage at http://localhost:8000/ with and without the flag to compare the number of database queries (151 vs 46).

You can further test both the existing and new backends with the two proposed mega menu designs by setting the `MEGA_MENU_VAR_1` and `MEGA_MENU_VAR_2` flags. In both cases, the design and behavior should be the same.

CFPB devs can also play with a version of this branch on our internal DEV4 server.

## Todos

This is a big change but there are several TODOs that I have kept out of this PR to keep its scope as small as possible:

- Update the documentation at http://cfpb.github.io/cfgov-refresh/mega-menu/ to reflect the new editing experience.
- Add better help text and form labels in the Wagtail menu editor.
- Create a nicer/more compact link chooser to reduce the duplication of (page, text, link) in the Wagtail menu editor (this might have to wait until we are on Wagtail 2+ to be able to leverage packages like [wagtail-generic-chooser](https://github.com/wagtail/wagtail-generic-chooser)).
- Move existing mega menu templates from jinja2/v1/_includes to the mega_menu app (if FEWDs agree this makes sense).
- Make the "Preview" button on the menu edit screen preview unsaved edits, the way the preview button on the Wagtail page editor works (requires some JS code).
- Add documentation of the expected menu content format to the frontend template.
- Remove legacy menu code and related infrastructure (the Django `default_fragment_cache` cache, which will no longer be needed after this change).

## Checklist
